File: SemanticScript/bench/algorithms/sieve.py
"""sieve.py - Sieve of Eratosthenes up to LIMIT, repeated REPEATS times.

Mirrors sieve.c's algorithm and checksum, written in idiomatic fast Python:
  - a bytearray holds 1 for "still prime", 0 for "composite";
  - composites are struck out with a single strided slice assignment per prime
    (a C-level bulk store) instead of a Python-level inner loop;
  - the outer loop only runs to sqrt(LIMIT), since every composite has a prime
    factor at or below its square root;
  - the prime count is sum() over the bytearray, another C-level pass.

This is the form a competent Python developer writes for speed. The task and
the resulting prime count are identical to the other three languages; only the
in-language mechanics differ. Timed with time.perf_counter().
"""

import time
from math import isqrt


def main():
    limit = 2000000
    repeats = 20
    array_size = limit + 1
    sqrt_limit = isqrt(limit)
    total_prime_count = 0

    start = time.perf_counter()
    for _ in range(repeats):
        sieve = bytearray([1]) * array_size
        sieve[0:2] = b"\x00\x00"
        for candidate in range(2, sqrt_limit + 1):
            if sieve[candidate]:
                first_composite = candidate * candidate
                composite_count = (limit - first_composite) // candidate + 1
                sieve[first_composite::candidate] = b"\x00" * composite_count
        prime_count = sum(sieve)
        total_prime_count += prime_count
    end = time.perf_counter()

    elapsed_seconds = end - start
    print(f"checksum={total_prime_count} elapsedSeconds={elapsed_seconds:.6f}")


if __name__ == "__main__":
    main()

File: legacy/SemanticScript/bench/algorithms/sieve.js
// sieve.js - Sieve of Eratosthenes up to LIMIT, repeated REPEATS times.
// Same structure as the other three languages: a Uint8Array starts all-prime
// (1), composites are struck out (0) for primes up to sqrt(LIMIT), and the
// prime count is a final pass. Wrapped in main() so V8 JIT-optimizes the
// hot loops. Same LIMIT/REPEATS, same checksum, same output contract.

function main() {
  const limit = 2000000;
  const repeats = 40;
  const arraySize = limit + 1;

  const sieve = new Uint8Array(arraySize);
  let totalPrimeCount = 0;

  const start = performance.now();
  for (let repeat = 0; repeat < repeats; repeat++) {
    sieve.fill(1);
    sieve[0] = 0;
    sieve[1] = 0;
    for (let candidate = 2; candidate * candidate <= limit; candidate++) {
      if (sieve[candidate]) {
        for (let multiple = candidate * candidate; multiple <= limit; multiple += candidate) {
          sieve[multiple] = 0;
        }
      }
    }
    let primeCount = 0;
    for (let index = 2; index <= limit; index++) {
      primeCount += sieve[index];
    }
    totalPrimeCount += primeCount;
  }
  const end = performance.now();

  const elapsedSeconds = (end - start) / 1000;
  console.log(`checksum=${totalPrimeCount} elapsedSeconds=${elapsedSeconds.toFixed(6)}`);
}

main();
